### ISSUE TITLE
Pop-up bookmarks/notes

### ISSUE THREAD

**Issue by Markismus** [koreader/koreader#6100]
I really like the ~in-page~pop-up footnotes. It's brilliant. Apparently, there is always some disorientation when the page changes to go to the footnote section and this prevents it.

I would also like to see the bookmarks/notes in such a way. So when I tap a highlight, I would get the content of the bookmark displayed in-page. Tapping that area could bring you to the editor, while holding could give a pop-up for deletion.

What parts of the code could I look at to try and hack this feature into Koreader? Is it a tweak or would it require a lot of coding?

**Comment by poire-z** [comment]
Not sure if you mean _in-pages footnotes_ with "in-page footnotes" :)

These are in-page footnotes:
<kbd>![image](https://user-images.githubusercontent.com/24273478/80461547-bcadcc00-8935-11ea-8530-dff66570c4b9.png)</kbd>

These are popup footnotes:
<kbd>![image](https://user-images.githubusercontent.com/24273478/80461592-cd5e4200-8935-11ea-950d-bdc068940f98.png)</kbd>

In-page footnotes are handled by crengine (the epub renderer), and there will be no way to have it know about bookmarks/highlight/notes (and keep track of, insert that as HTML content, render it...) - which are handled by KOReader only. (and that would need a re-rendering of the whole book each time you modify a note).

popup footnotes are handled and drawn by KOReader, as a layer over the book rendered content.
So, I guess showing your notes for a bookmark this way (so, after you tap on it) would be feasible.
The code that handles that is at the bottom of frontend/apps/reader/modules/readerlink.lua and frontend/ui/widget/footnotewidget.lua which is mostly dealing with having MuPDF drawing HMTL snippets.

But I'm not so sure it's a good idea :)
- most people don't add notes - so they would see the exact same highlighted text in that popup
- and if you do add notes, and the highlighted text is at the bottom of the page, the popup footnotes would cover the original text, and you might be with your notes without its context. (that was easier to ignore for footnotes, which are, well, at the foot of the page :)

So, may be you'd just want to rework the show/edit widget for highlights:

But you'd have to think about what to do with the initial button dialog, that I guess is more useful to most people than just going directly to (re-)show the highlighted text and notes:
<kbd>![image](https://user-images.githubusercontent.com/24273478/80462811-8f621d80-8937-11ea-8cc3-985e63adeb6a.png)</kbd>

> What parts of the code could I look at to try and hack this feature into Koreader? Is it a tweak or would it require a lot of coding?

Would probably require much coding in frontend/apps/reader/modules/readerhighlight.lua or readerbookmark.lua.

Also note #5794 (a bit stalled), after which we might know more about highlights that have notes and those that don't.

**Comment by Markismus** [comment]
Thanks @poire-z! I do agree that it should be a toggle: Just like in-page footnotes shown as pop-ups in style tweaks>Miscellaneous>In-page footnotes>In page Epub footnotes(smaller).

Does the show/edit widget for highlights prevent obscuring the highlighted text?

**Comment by poire-z** [comment]
> Just like in-page footnotes shown as pop-ups in style tweaks>Miscellaneous>In-page footnotes>In page Epub footnotes(smaller).

You're mixing in-page footnotes & popup footnotes: they are independant thingies, you can enable one or the other or both, but they are unrelated.
In-page footnotes (the style twaks you mentionned) are not shown as pop-ups.
Popup footnotes (Gear > Taps and gestures > Links > Show footnotes in popup) are... shown as popups :)

> Does the show/edit widget for highlights prevent obscuring the highlighted text?

?
Currently, the button dialog is displayed in the center - and the edit, showing the keyboard, will/might all be over the highlighted text.

**Comment by Markismus** [comment]
Yes! You are perfectly right. I did have the wrong setting. I meant the feature enabled with Gear > Taps and gestures > Links > Show footnotes in popup.

**Comment by poire-z** [comment]
Well, in that case you could add an option (in Gear>Document>, where there are already a few things related to highlights), and instead of showing that `delete|edit|...`, you would show some kind of new widget with your notes, and try to finely position it depending on highlight position in page, the space available above or below, the height of the note text... (we do that for the dictionary popup, but it works only when the selected word is really at top or at bottom of page).

But you'll have to really get into KOReader widget/UI/event stuff (you'd have quite a few existing widgets to take inspiration from - dictwidget, popupwidget,... but it's still quite something to get into).

**Comment by Markismus** [comment]
That is indeed not very encouraging.

**Comment by gnoownow10** [comment]
How about changes like this? I have no prior experience in lua, so I don't know this is good, but this serves me well, at least for my usage...

~~~patch
diff --git a/frontend/apps/reader/modules/readerhighlight.lua b/frontend/apps/reader/modules/readerhighlight.lua
index 9fbaf0c7..de7a74a0 100644
--- a/frontend/apps/reader/modules/readerhighlight.lua
+++ b/frontend/apps/reader/modules/readerhighlight.lua
@@ -1,5 +1,6 @@
 local BD = require("ui/bidi")
 local ButtonDialog = require("ui/widget/buttondialog")
+local ButtonDialogTitle = require("ui/widget/buttondialogtitle")
 local Device = require("device")
 local Event = require("ui/event")
 local InfoMessage = require("ui/widget/infomessage")
@@ -686,9 +687,29 @@ function ReaderHighlight:onShowHighlightDialog(page, index)
             }
         })
     end
-    self.edit_highlight_dialog = ButtonDialog:new{
-        buttons = buttons
-    }
+
+    local item = self.view.highlight.saved[page][index]
+    local found = false
+    for __, bookmark in ipairs(self.ui.bookmark.bookmarks) do
+        if bookmark.highlighted then
+            if bookmark.pos0 == item.pos0 and bookmark.pos1 == item.pos1 then
+                if bookmark.text and bookmark.text ~= "" then
+                    self.edit_highlight_dialog = ButtonDialogTitle:new{
+                        buttons = buttons,
+                        title = bookmark.text
+                    }
+                    found = true
+                    break
+                end
+            end
+        end
+    end
+    if not found then
+        self.edit_highlight_dialog = ButtonDialog:new{
+            buttons = buttons
+        }
+    end
+
     UIManager:show(self.edit_highlight_dialog)
     return true
 end
~~~

**Comment by poire-z** [comment]
But if you haven't modified the text (= adding a note), you'll get in the widget the exact same text that you are already seeing highlighted, which wouldn't be useful and would just be noise.
We'd need to know it's a note (or that it is different from the source text), and this would need us to revamp the whole bookmark/highlight to make it a bit more rational: #5794. For now, it's just a mess and everything would just be more hacks around the mess :/

**Comment by poire-z** [comment]
> I would also like to see the bookmarks/notes in such a way. So when I tap a highlight, I would get the content of the bookmark displayed in-page. Tapping that area could bring you to the editor, while holding could give a pop-up for deletion.

Implemented by #8600 :
![06](https://user-images.githubusercontent.com/62179190/147660936-07175f16-124c-4d77-8c9f-d9b2dbe18497.png)
